pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -186,6 +216,240 @@ func (r *Registry) Remove(id ipc.ID, creds *auth.Credentials) error {
 	return nil
 }
 
+// FindByID returns the queue with the specified ID and an error if the ID
+// doesn't exist.
+func (r *Registry) FindByID(id ipc.ID) (*Queue, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	mech := r.reg.FindByID(id)
+	if mech == nil {
+		return nil, syserror.EINVAL
+	}
+	return mech.(*Queue), nil
+}
+
+// Send appends a message to the message queue, and returns an error if sending
+// fails. See msgsnd(2).
+func (q *Queue) Send(ctx context.Context, m Message, b Blocker, wait bool, pid int32) error {
+	creds := auth.CredentialsFromContext(ctx)
+	if err := q.append(ctx, m, creds, pid); err != syserror.ErrWouldBlock {
+		return err
+	}
+
+	if !wait {
+		return syserror.EAGAIN
+	}
+
+	e, ch := waiter.NewChannelEntry(nil)
+	q.senders.EventRegister(&e, waiter.EventOut)
+
+	err := q.append(ctx, m, creds, pid)
+	for ; err == syserror.ErrWouldBlock; err = q.append(ctx, m, creds, pid) {
```

Comment:
If someone tries to queue a message larger than the max queue size, doesn't this loop forever?

Reply:
Messages have a maximum size `linux.MSGMAX`, which we check in `Msgsnd`. The reason for that is that `msgsnd` receives `msgsz` as an input, so we want to check if the size is valid before copying in data.

Also, the max message size `linux.MSGMAX` is lower than the max queue size `linux.MSGMNB`, which means that we always have enough space for at least one message in the queue. So theoretically, the loop will not run forever.